### PR TITLE
perf(core): improve the internal Worker pool

### PR DESCRIPTION
**Pull request by merceyz**
**What's the problem this PR addresses?**

- The Worker pool attaches new event listeners for every task assigned to it instead of reusing them.
- The Worker pool uses `p-queue` to control concurrency instead of `p-limit` which is used everywhere else.
- If the Worker pool is idle for about one second all Workers are terminated, this is problematic since it's quite likely for the pool to be idle for a few seconds while Yarn downloads new packages.

**How did you fix it?**

- Re-use the event listeners.
- Replace `p-queue` with `p-limit` reducing the bundle size by 8kB.
- Check if the pool is idle every five seconds, if it is idle, terminate one Worker and repeat.
When testing on the Storybook repo this reduces the amount of Workers created from 24 to 16 (the amount of cores on my machine).

**Result**

The runtime impact on the gatsby benchmark is low enough to be written off as noise but including it anyways
```sh
$ YARN_IGNORE_PATH=1 hyperfine --prepare="rm -rf ./.yarn/global" -w 1 "node ./before.cjs --mode skip-build" "node ./after.cjs --mode skip-build" "node ./plimit.cjs --mode skip-build"
Benchmark #1: node ./before.cjs --mode skip-build
  Time (mean ± σ):      8.402 s ±  0.075 s    [User: 52.890 s, System: 3.390 s]
  Range (min … max):    8.269 s …  8.510 s    10 runs

Benchmark #2: node ./after.cjs --mode skip-build
  Time (mean ± σ):      8.397 s ±  0.070 s    [User: 52.964 s, System: 3.394 s]
  Range (min … max):    8.296 s …  8.494 s    10 runs

Benchmark #3: node ./plimit.cjs --mode skip-build
  Time (mean ± σ):      8.396 s ±  0.071 s    [User: 53.030 s, System: 3.448 s]
  Range (min … max):    8.281 s …  8.487 s    10 runs

Summary
  'node ./plimit.cjs --mode skip-build' ran
    1.00 ± 0.01 times faster than 'node ./after.cjs --mode skip-build'
    1.00 ± 0.01 times faster than 'node ./before.cjs --mode skip-build'
```

**Checklist**
- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).
- [x] I have set the packages that need to be released for my changes to be effective.
- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/yarnpkg-core/sources/WorkerPool.ts [hunk 1]
```diff
@@ -1,53 +1,67 @@
...
-  });
+  private workers: Array<PoolWorker<TOut>> = [];
+  private limit = PLimit(Math.max(1, cpus().length));
 
   constructor(private source: string) {
-    const timeout = setTimeout(() => {
-      if (this.queue.size !== 0 || this.queue.pending !== 0)
-        return;
+    setInterval(() => {
+      if (this.limit.pendingCount == 0 && this.limit.activeCount === 0) {
```
Comment (paul-soporan) [inline packages/yarnpkg-core/sources/WorkerPool.ts:17]: Is there a reason for not using strict equality with `pendingCount` (which seems to be a `number` according to the types)?

Comment (merceyz) [reply packages/yarnpkg-core/sources/WorkerPool.ts]: No, that was not done on purpose, nice catch 👍

File: packages/yarnpkg-core/sources/WorkerPool.ts [hunk 2]
```diff
@@ -1,53 +1,67 @@
...
-    concurrency: Math.max(1, cpus().length),
-  });
+  private workers: Array<PoolWorker<TOut>> = [];
+  private limit = PLimit(Math.max(1, cpus().length));
 
   constructor(private source: string) {
-    const timeout = setTimeout(() => {
-      if (this.queue.size !== 0 || this.queue.pending !== 0)
-        return;
+    setInterval(() => {
```
Comment (arcanis) [inline packages/yarnpkg-core/sources/WorkerPool.ts:16]: As is, this setInterval will keep executing even after all the processes have been killed (vs the previous setTimeout-based implementation, which stops once there's nothing to cleanup).

**Review by arcanis — COMMENTED** [review]
Just one detail about the cleanup interval

File: packages/yarnpkg-core/sources/WorkerPool.ts (hunk 2 above)
Comment (merceyz) [reply packages/yarnpkg-core/sources/WorkerPool.ts]: Yeah, since it's `unref`ed I didn't think much of it but I agree, it should clean up after itself 👍

**Pull request merged**